flight/Modules/CharacterOSD/panel.c

```diff
@@ -0,0 +1,471 @@
+	}
+
+	draw_rect(state, x, y, 6, 3, 0, 0);
+	PIOS_MAX7456_puts(state->dev, x + 1, y + 1, mode, 0);
+}
+
+/* ArmedFlag */
+static void ARMEDFLAG_update(charosd_state_t state, uint8_t x, uint8_t y)
+{
+	uint8_t attr = state->telemetry.flight_status.arm_status == FLIGHTSTATUS_ARMED_ARMED ? 0 : MAX7456_ATTR_INVERT;
+	draw_rect(state, x, y, 3, 3, true, attr);
+	PIOS_MAX7456_put(state->dev, x + 1, y + 1, 0xe0, attr);
+}
+
+static void FLIGHTTIME_update(charosd_state_t state, uint8_t x, uint8_t y) {
+       char buffer[10];
+       uint32_t time = state->telemetry.system.flight_time;
+       int min, sec;
+
+       uint16_t hours = (time / 3600000); // hours
+       if (hours == 0) {
+               min = time / 60000;
+               sec = (time / 1000) - 60 * min;
+               sprintf(buffer, "%02d:%02d", (int)min, (int)sec);
+       } else {
+               min = time / 60000 - 60 * hours;
+               sec = (time / 1000) - 60 * min - 3600 * hours;
+               sprintf(buffer, "%02d:%02d:%02d", (int)hours, (int)min, (int)sec);
+       }
+
+       terminate_buffer();
+
+       PIOS_MAX7456_puts(state->dev, x + 1, y + 1, buffer, 0);
+}
+
+/* Roll */
+STD_PANEL(ROLL, 7, "\xb2%d\xb0", (int16_t) state->telemetry.attitude_actual.Roll);
+
+/* Pitch */
+STD_PANEL(PITCH, 7, "\xb1%d\xb0", (int16_t) state->telemetry.attitude_actual.Pitch);
+
+/* GPS */
+
+#define _PAN_GPS_2D 0x01
+#define _PAN_GPS_3D 0x02
+
+static void GPS_update(charosd_state_t state, uint8_t x, uint8_t y)
+{
+	char buffer[4];
+
+	snprintf(buffer, sizeof(buffer), "%d", state->telemetry.gps_position.Satellites); 
+	terminate_buffer();
+	bool err = state->telemetry.gps_position.Status < GPSPOSITION_STATUS_FIX2D;
+	PIOS_MAX7456_puts(state->dev, x, y, "\x10\x11", err ? MAX7456_ATTR_INVERT : 0);
+	PIOS_MAX7456_put(state->dev, x + 2, y, state->telemetry.gps_position.Status < GPSPOSITION_STATUS_FIX3D ? _PAN_GPS_2D : _PAN_GPS_3D,
+		err ? MAX7456_ATTR_INVERT : (state->telemetry.gps_position.Status < GPSPOSITION_STATUS_FIX2D ? MAX7456_ATTR_BLINK : 0));
+	if (err) PIOS_MAX7456_puts (state->dev, x + 3, y, ERR_STR, MAX7456_ATTR_INVERT);
+	else PIOS_MAX7456_puts(state->dev, x + 3, y, buffer, 0);
+}
+
+/* Lat */
+STD_PANEL(LATITUDE, 11, "\x83%02.6f", (double)state->telemetry.gps_position.Latitude / 10000000.0);
+
+/* Lon */
+STD_PANEL(LONGITUDE, 11, "\x84%02.6f", (double)state->telemetry.gps_position.Longitude / 10000000.0);
+
+#define PANEL_HORIZON_WIDTH 14
+#define PANEL_HORIZON_HEIGHT 5
+
+#define _PAN_HORZ_LINE 0x16
+#define _PAN_HORZ_TOP 0x0e
+
+#define _PAN_HORZ_CHAR_LINES 18
+#define _PAN_HORZ_VRES 9
+#define _PAN_HORZ_INT_WIDTH (PANEL_HORIZON_WIDTH - 2)
+#define _PAN_HORZ_LINES (PANEL_HORIZON_HEIGHT * _PAN_HORZ_VRES)
+#define _PAN_HORZ_TOTAL_LINES (PANEL_HORIZON_HEIGHT * _PAN_HORZ_CHAR_LINES)
+
+const char _horz_line [PANEL_HORIZON_WIDTH + 1]   = "\xb8            \xb9";
+const char _horz_center [PANEL_HORIZON_WIDTH + 1] = "\xc8            \xc9";
+static double deg2rad(double deg) {
```

Comment:
miscmath.h's DEG2RAD is a constant.  Also can these be floats instead of double?